### PR TITLE
PIM-5422: fix memory leak on import products

### PR DESCRIPTION
**Pull request by skeleton**
| Q                 | A
| ----------------- | ---
| Added Specs       | N
| Added Behats      | N
| Travis CI is ok   | Y
| Changelog updated | Y

A cherry pick has been done from 1.4 to detach product value.
UniqueVariantAxisValidator has been reworked to avoid to hydrate products whereas we need only the id.

Result, import a rule with 3000 products inpacted:

before patch | after patch
------------ | -------------
20 minutes | 13 minutes
625Mb | 70.75Mb

### CODE REVIEW

File: src/Akeneo/Bundle/StorageUtilsBundle/Doctrine/Common/Detacher/ObjectDetacher.php [hunk 1]
```diff
@@ -33,7 +36,13 @@ public function __construct(ManagerRegistry $registry)
     public function detach($object)
     {
         $objectManager = $this->getObjectManager($object);
-        $objectManager->detach($object);
+
+        if ($objectManager instanceof DocumentManager) {
+            $visited = [];
```
Comment (grena) [inline src/Akeneo/Bundle/StorageUtilsBundle/Doctrine/Common/Detacher/ObjectDetacher.php]: Can `$visited` change? passed by reference in the method below?

File: src/Akeneo/Bundle/StorageUtilsBundle/Doctrine/Common/Detacher/ObjectDetacher.php [hunk 2]
```diff
@@ -45,4 +54,57 @@ protected function getObjectManager($object)
...
+            }
+            $relatedDocuments = $class->reflFields[$mapping['fieldName']]->getValue($document);
+            if (($relatedDocuments instanceof Collection || is_array($relatedDocuments))) {
+                if ($relatedDocuments instanceof PersistentCollection) {
+                    $relatedDocuments = $relatedDocuments->unwrap();
+                }
+                foreach ($relatedDocuments as $relatedDocument) {
+                    $this->doDetach($relatedDocument, $visited);
+                }
+            } elseif ($relatedDocuments !== null) {
```
Comment (solivier) [inline src/Akeneo/Bundle/StorageUtilsBundle/Doctrine/Common/Detacher/ObjectDetacher.php]: yoda style

File: src/Akeneo/Bundle/StorageUtilsBundle/Doctrine/Common/Detacher/ObjectDetacher.php [hunk 3]
```diff
@@ -45,4 +54,57 @@ protected function getObjectManager($object)
...
+     * @param array  $visited Prevent infinite recursion
+     */
+    protected function cascadeDetach($document, array &$visited)
+    {
+        $documentManager = $this->getObjectManager($document);
+
+        $class = $documentManager->getClassMetadata(ClassUtils::getClass($document));
+        foreach ($class->fieldMappings as $mapping) {
+            if (!$mapping['isCascadeDetach']) {
+                continue;
```
Comment (solivier) [inline src/Akeneo/Bundle/StorageUtilsBundle/Doctrine/Common/Detacher/ObjectDetacher.php]: Not of big fan of continue in a foreach.

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/ProductRepository.php [hunk 4]
```diff
@@ -746,4 +733,52 @@ public function getProductCountByGroup(GroupInterface $group)
...
+    {
+        $qb = $this->findAllForVariantGroupQB($variantGroup, $criteria);
+        $qb
+            ->select('_id')
+            ->hydrate(false);
+
+        $cursor = $qb->getQuery()->execute();
+
+        $products = [];
+        foreach ($cursor as $product) {
```
Comment (solivier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/ProductRepository.php]: why using a cursor here as you fetch all the VariantGroup in a foreach ?

**Comment by solivier** [comment]
GTM

**Comment by grena** [comment]
GTM

**Comment by jjanvier** [comment]
gtm

**Pull request merged**